From 727b7e86afe3acbe231974033e08e765b5354614 Mon Sep 17 00:00:00 2001
From: Paul Praschl <maintainer@example.invalid>
Date: Tue, 14 Jul 2026 13:05:53 +0200
Subject: [PATCH 04/43] arm64: dts: apple: Add T8010 PCIe and NVMe

Add the T8010 PCIe host bridge and storage-port DART to the SoC
description. Enable them for the common T8010 platform and reserve an
empty NVMe HMB node for the loader to populate from the runtime ADT.

Describe the storage endpoint as root port 0 and route its requester ID
through the matching DART stream.

Signed-off-by: Paul Praschl <maintainer@example.invalid>
(cherry picked from commit d7c63aedb9a7c80690e4500590e0cd96bfdeb400)
---
 arch/arm64/boot/dts/apple/t8010-common.dtsi | 14 +++-
 arch/arm64/boot/dts/apple/t8010.dtsi        | 72 +++++++++++++++++++++
 2 files changed, 85 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/apple/t8010-common.dtsi b/arch/arm64/boot/dts/apple/t8010-common.dtsi
index 7426147af3..e583b1c576 100644
--- a/arch/arm64/boot/dts/apple/t8010-common.dtsi
+++ b/arch/arm64/boot/dts/apple/t8010-common.dtsi
@@ -12,6 +12,7 @@
 / {
 	aliases {
 		ausbphy0 = &ausbphy0; /* Used by loader */
+		nvme-hmb = &nvme_hmb; /* Used by loader */
 		serial0 = &serial0;
 	};
 
@@ -40,7 +41,10 @@ reserved-memory {
 		#size-cells = <2>;
 		ranges;
 
-		/* To be filled by loader */
+		nvme_hmb: nvme-hmb@0 {
+			reg = <0 0 0 0>; /* To be filled by loader */
+			no-map;
+		};
 	};
 };
 
@@ -48,6 +52,14 @@ &dwi_bl {
 	status = "okay";
 };
 
+&pcie0_dart0 {
+	status = "okay";
+};
+
+&pcie0 {
+	status = "okay";
+};
+
 &serial0 {
 	status = "okay";
 };
diff --git a/arch/arm64/boot/dts/apple/t8010.dtsi b/arch/arm64/boot/dts/apple/t8010.dtsi
index 72ec543d7c..e77f5d28ad 100644
--- a/arch/arm64/boot/dts/apple/t8010.dtsi
+++ b/arch/arm64/boot/dts/apple/t8010.dtsi
@@ -276,6 +276,78 @@ ehci1: usb-ehci@20c400000 {
 #endif
 		};
 
+		pcie0_dart0: iommu@601008000 {
+			compatible = "apple,t8010-dart", "apple,s5l8960x-dart";
+			reg = <0x6 0x01008000 0x0 0x4000>;
+			#iommu-cells = <1>;
+			interrupts = <AIC_IRQ 271 IRQ_TYPE_LEVEL_HIGH>;
+			power-domains = <&ps_pcie>;
+			apple,dma-range = <0x0 0x80000000 0x0 0x3c000000>;
+			status = "disabled";
+		};
+
+		pcie0: pcie@610000000 {
+			compatible = "apple,t8010-pcie";
+			device_type = "pci";
+
+			reg = <0x6 0x10000000 0x0 0x1000000>,
+			      <0x6 0x00000000 0x0 0x8000>,
+			      <0x6 0x00008000 0x0 0x4000>,
+			      <0x6 0x0a000000 0x0 0x40000>,
+			      <0x6 0x01000000 0x0 0x4000>,
+			      <0x6 0x02000000 0x0 0x4000>,
+			      <0x6 0x03000000 0x0 0x4000>,
+			      <0x6 0x04000000 0x0 0x4000>,
+			      <0x6 0x01004000 0x0 0x4000>;
+			reg-names = "config", "rc", "phy", "phy-ip",
+				    "port0", "port1", "port2", "port3",
+				    "nvmmu";
+
+			interrupts = <AIC_IRQ 270 IRQ_TYPE_LEVEL_HIGH>,
+				     <AIC_IRQ 273 IRQ_TYPE_LEVEL_HIGH>,
+				     <AIC_IRQ 276 IRQ_TYPE_LEVEL_HIGH>,
+				     <AIC_IRQ 279 IRQ_TYPE_LEVEL_HIGH>,
+				     <AIC_IRQ 272 IRQ_TYPE_EDGE_RISING>;
+			interrupt-names = "port0", "port1", "port2", "port3",
+					  "nvmmu";
+
+			msi-controller;
+			msi-parent = <&pcie0>;
+			msi-ranges = <&aic AIC_IRQ 288 IRQ_TYPE_EDGE_RISING 32>;
+			apple,msi-doorbell = <0xbffff000>;
+			apple,nvmmu-iova = <0xbc000000>;
+			apple,nvme-hmb = <&nvme_hmb>;
+
+			iommu-map = <0x100 &pcie0_dart0 0 1>;
+			iommu-map-mask = <0xff00>;
+
+			#address-cells = <3>;
+			#size-cells = <2>;
+			#interrupt-cells = <1>;
+			interrupt-controller;
+			bus-range = <0x00 0x0f>;
+			ranges = <0x03000000 0x0 0xc0000000
+				  0x7 0xc0000000 0x0 0x40000000>;
+
+			clocks = <&clkref>, <&clkref>, <&clkref>;
+			clock-names = "core", "aux", "ref";
+			power-domains = <&ps_pcie>, <&ps_pcie_aux>,
+					<&ps_pcie_ref>;
+			power-domain-names = "core", "aux", "ref";
+			status = "disabled";
+
+			pci@0,0 {
+				device_type = "pci";
+				reg = <0x0 0x0 0x0 0x0 0x0>;
+				reset-gpios = <&pinctrl_ap 12 GPIO_ACTIVE_LOW>;
+				clkreq-gpios = <&pinctrl_ap 16 GPIO_ACTIVE_LOW>;
+
+				#address-cells = <3>;
+				#size-cells = <2>;
+				ranges;
+			};
+		};
+
 		pmgr: power-management@20e000000 {
 			compatible = "apple,t8010-pmgr", "apple,pmgr", "syscon", "simple-mfd";
 			#address-cells = <1>;
-- 
2.56.0

